subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/IncrementalTaskAction.java

```diff
@@ -54,39 +51,29 @@ public void releaseContext() {
-                    return changes.getInputFilesChanges().map(new Function<Iterable<Change>, StatefulIncrementalTaskInputs>() {
-                        @Override
-                        public StatefulIncrementalTaskInputs apply(Iterable<Change> changes) {
-                            return createIncrementalInputs(changes);
-                        }
-                    }).orElseGet(new Supplier<StatefulIncrementalTaskInputs>() {
-                        @Override
-                        public StatefulIncrementalTaskInputs get() {
-                            return createRebuildInputs(task);
-                        }
-                    });
-                }
-            }).orElseGet(new Supplier<StatefulIncrementalTaskInputs>() {
-                @Override
-                public StatefulIncrementalTaskInputs get() {
-                    return createRebuildInputs(task);
-                }
-            });
+        @SuppressWarnings("OptionalGetWithoutIsPresent")
+        ExecutionStateChanges changes = context.getExecutionStateChanges().get();
+        IncrementalTaskInputs incrementalTaskInputs = changes.visitInputFileChanges(new ExecutionStateChanges.IncrementalInputsVisitor<IncrementalTaskInputs>() {
+            @Override
+            public IncrementalTaskInputs visitRebuild(ImmutableSortedMap<String, CurrentFileCollectionFingerprint> allFileInputs) {
+                return createRebuildInputs(task, allFileInputs);
+            }
 
-        context.setTaskExecutedIncrementally(incrementalInputs.isIncremental());
-        JavaMethod.of(task, Object.class, methodName, IncrementalTaskInputs.class).invoke(task, incrementalInputs);
+            @Override
+            public IncrementalTaskInputs visitIncrementalChange(Iterable<Change> inputFileChanges) {
+                return createIncrementalInputs(inputFileChanges);
+            }
+        });
+
+        context.setTaskExecutedIncrementally(incrementalTaskInputs.isIncremental());
+        JavaMethod.of(task, Object.class, methodName, IncrementalTaskInputs.class).invoke(task, incrementalTaskInputs);
     }
 
     private ChangesOnlyIncrementalTaskInputs createIncrementalInputs(Iterable<Change> inputFilesChanges) {
         return instantiator.newInstance(ChangesOnlyIncrementalTaskInputs.class, inputFilesChanges);
     }
 
-    private RebuildIncrementalTaskInputs createRebuildInputs(Task task) {
-        ImmutableCollection<CurrentFileCollectionFingerprint> currentInputs = context.getBeforeExecutionState().get().getInputFileProperties().values();
-        return instantiator.newInstance(RebuildIncrementalTaskInputs.class, task, currentInputs);
+    private RebuildIncrementalTaskInputs createRebuildInputs(Task task, ImmutableSortedMap<String, CurrentFileCollectionFingerprint> currentInputs) {
+        return instantiator.newInstance(RebuildIncrementalTaskInputs.class, task, currentInputs.values());
```

Comment:
The only use of `task` here is to print this:

```java
        LOGGER.info("All input files are considered out-of-date for incremental {}.", task);
```

We should not pass it around as much, and we can potentially emit this log message in the execution engine, which would be better.

Reply:
Agreed. I didn't want to change that here, yet.